### DISCUSSION TITLE
How to translate path in URL?

### DISCUSSION THREAD

**Discussion by polenter** [vuejs/vitepress#2857] [Q&A] [answered]
Is there any possibility to define a target page for the "language selector" button, individually for each page?

There are two pages, concerning the same subject, one in EN, the other in DE.
The structure is as follows:
```
/cookbook
/de/kochbuch
```

The link from `/cookbook` is `/de/cookbook`,
however expected is `/de/kochbuch`.

Is there any possibility to configure it in VitePress?

**Comment by brc-dd** [comment, marked as answer]
You'll need to patch `normalizeLink` in `dist/client/theme-default/composables/langs.js` for this. Other than that there is currently no way 👀

**Reply by polenter** [reply]
First, thanks for the prompt response.

There is no `client/` and no `langs.js` in my `dist` folder.

But you mean text find/replace after generating the output, don't you?
That could work, however it's somewhat a brute force.

So there is no hook, no event, no watcher, no mapping, no rewrite etc...

I miss this feature so much.
Translating paths improves SEO. 
Furthermore, the path does not scarry the visitor.

**Reply by brc-dd** [reply]
Ah, no. Not your dist output. Patch vitepress: https://pnpm.io/cli/patch You can also do this: https://vitepress.dev/guide/extending-default-theme#overriding-internal-components

**Reply by polenter** [reply]
Thank you for the link to `pnpm patch`. It has inspired me.  
I've patched `vitepress` using `npm patch-package`, as in the following link.  
https://www.npmjs.com/package/patch-package

Patched was `node_modules/vitepress/dist/client/theme-default/composables/langs.js`

Get `frontmatter` object from `useData`.
```
const { site, localeIndex, page, theme, frontmatter } = useData()
```

Insert `frontmatter.value.locales?.[key]?.link` for generating `path` from frontmatter.  
Is there no `locales` in frontmatter defined, the default link will be generated, based on the site configuration.

```
frontmatter.value.locales?.[key]?.link ?? page.value.relativePath.slice(currentLang.value.link.length - 1),
```

Define link in the frontmatter of the `root `page.  
There can be as many links, as many languages defined.

`/docs/welcome.md`
```
---
locales: {
  de: {
    link: "/docs/willkommen"
  },
  fr: {
    link: "/docs/bienvenue"
  }
}
---
```

Define links in frontmatter of the translated pages.  
Following is a german page.

`/de/docs/willkommen.md`
```
---
locales: {
  root: {
    link: "/docs/welcome"
  },
  fr: {
    link: "/docs/bienvenue"
  }
}
---
```

If the page structure is in format `/docs/welcome/index.md`, and not `/docs/welcome.md`, each link must end with `/`.

`link: "/docs/welcome/"`

If there is demand, I could provide a PR with the above patch.

Thanks again for your help.
